### PR TITLE
Visibility sim overhaul

### PR DESCRIPTION
**Pull request by steven-murray**
This is the main conglomerating hub for the visibility simulator overhaul.

### CODE REVIEW

**Comment by steven-murray** [comment]
@philbull I've attempted to merge in the polybeam stuff from master back into this PR. A lot of it was easy, but there were some conflicts I wasn't quite sure of. I'd be grateful if you could take a quick look over the relevant files!

**Comment by steven-murray** [comment]
TODO:

- [] Tutorial notebooks (port over from end-to-end example)
- [] Ensure tests are really testing stuff. Eg. test that specifying multiple beams gets used appropriately, polarization consistency checks

**Comment by piyanatk** [comment]
Also maybe be more specific here that `sky_model` is a `` :class:~`pyradiosky.SkyModel` ``?

https://github.com/HERA-Team/hera_sim/blob/ab5d8e14cad3a730b8cd1d0e1c0dec443532f9d8/hera_sim/visibilities/simulators.py#L36-L37

File: hera_sim/visibilities/vis_cpu.py [hunk 1]
```diff
@@ -316,58 +269,71 @@ def _base_simulate(self, crd_eq, I_sky):
             myid = self.mpi_comm.Get_rank()
             nproc = self.mpi_comm.Get_size()
 
+        if self.correct_source_positions:
+            # TODO: check if this is the right time to be using...
+            ra, dec = self.correct_point_source_pos(
+                data_model, obstime=data_model.uvdata.time_array[0]
+            )
```
Comment (piyanatk) [inline hera_sim/visibilities/vis_cpu.py:276]: @steven-murray This will fail because `uvdata.time_array` is a float but `correct_point_source_pos` expect a string or a Quantity.

**Comment by steven-murray** [comment]
@piyanatk the `SkyModel` class is explicitly documented by its typing annotation.

File: hera_sim/visibilities/vis_cpu.py [hunk 2]
```diff
@@ -45,22 +45,35 @@ class VisCPU(VisibilitySimulator):
...
-        use_gpu=False,
+        bm_pix: int = 100,
+        use_pixel_beams: bool = True,
+        polarized: bool = False,
+        precision: int = 1,
+        use_gpu: bool = False,
         mpi_comm=None,
-        **kwargs
+        ref_time: Optional[Union[str, Time]] = None,
+        correct_source_positions: bool | None = None,
```
Comment (piyanatk) [inline hera_sim/visibilities/vis_cpu.py:73]: `X | Y` typing is Python 3.10 syntax. The current setup config support `3.6`, `3.7`, and `3.8`

Comment (piyanatk) [reply hera_sim/visibilities/vis_cpu.py:73]: See last line at https://docs.python.org/3.10/library/typing.html#typing.Union

Comment (piyanatk) [reply hera_sim/visibilities/vis_cpu.py:73]: It might be more clear to just do `correct_source_positions: bool = False`

File: hera_sim/visibilities/vis_cpu.py [hunk 3]
```diff
@@ -45,22 +45,35 @@ class VisCPU(VisibilitySimulator):
...
-        precision=1,
-        use_gpu=False,
+        bm_pix: int = 100,
+        use_pixel_beams: bool = True,
+        polarized: bool = False,
+        precision: int = 1,
+        use_gpu: bool = False,
         mpi_comm=None,
-        **kwargs
+        ref_time: Optional[Union[str, Time]] = None,
```
Comment (piyanatk) [inline hera_sim/visibilities/vis_cpu.py:72]: From the doc https://docs.python.org/3.10/library/typing.html#typing.Optional

`Optional[X]` is equivalent to `X | None` (or `Union[X, None]`).

So `ref_time: Optional[Union[str, Time]] = None` means `ref_time` can be type `Union[str, Time, None]`. The `Optional` qualifier seems irrelevant here.

**Review by piyanatk — CHANGES_REQUESTED** [review]
Some typing issues. One conflicts with `python>=3.6,<=3.8` requirement.

File: hera_sim/visibilities/vis_cpu.py (hunk 3 above)
Comment (steven-murray) [reply hera_sim/visibilities/vis_cpu.py:72]: I don't see how it's irrelevant? One of the options is None, so it definitely needs the `Optional`?

File: hera_sim/visibilities/vis_cpu.py [hunk 4]
```diff
@@ -316,58 +279,71 @@ def _base_simulate(self, crd_eq, I_sky):
...
+                    data_model.uvdata.antenna_numbers, data_model.uvdata.antenna_names
+                )
+                if number in ant_list
             ]
 
+        # Get x_orientation
+        x_orient = data_model.uvdata.x_orientation
+        if x_orient is None:
+            data_model.uvdata.x_orientation = "e"  # set in UVData object
+            x_orient = "e"  # default to east
```
Comment (piyanatk) [inline hera_sim/visibilities/vis_cpu.py:318]: `x_orientation` has to be either `"east"` or `"north"` (full word). See bottom of https://pyuvdata.readthedocs.io/en/latest/uvdata_parameters.html

**Comment by steven-murray** [comment]
@piyanatk and @r-pascua I've now merged #175. Please take another look and let's summarize what still needs to be done, if anything.

File: hera_sim/visibilities/vis_cpu.py (hunk 3 above)
Comment (r-pascua) [reply hera_sim/visibilities/vis_cpu.py:84]: In case you're still confused, I think that the value of the `Optional` type is its readability. Typing something as `Union[..., None]` doesn't make it quite as clear that the parameter is optional (i.e. doesn't need to be specified when calling the method/function) as the `Optional[...]` typing makes it.

Also, setting the default parameter value to `None` doesn't always mean that it's an optional parameter. Maybe it's not best practice to do this, but I've used it at least once in `hera_sim`: the `foregrounds.DiffuseForeground.__call__` method has the `Tsky` default set to `None`, and it will raise an error if it is not specified.

File: hera_sim/visibilities/healvis_wrapper.py [hunk 5]
```diff
@@ -36,64 +38,51 @@ class HealVis(VisibilitySimulator):
...
+        In addition to standard parameter restrictions, HealVis requires a single beam
+        for all antennae.
+        """
+        if model_data.n_beams > 1:
+            raise ValueError("healvis must use the same beam for all antennas.")
 
         # Check if pyuvsim.analyticbeam and switch to healvis.beam_model
-        if isinstance(self.beams[0], pyuvsim.analyticbeam.AnalyticBeam):
-            old_args = self.beams[0].__dict__
+        # TODO: we shouldn't silently modify model_data.beams...
```
Comment (r-pascua) [inline hera_sim/visibilities/healvis_wrapper.py:62]: Can we add a warning here so we're not silently modifying the beams?

File: hera_sim/visibilities/healvis_wrapper.py [hunk 6]
```diff
@@ -103,25 +92,34 @@ def sky_model(self):
...
+            conversion = (
+                1e-26
+                * sky_model.stokes.unit.to(units.Jy / units.sr)
+                * (cnst.c.si.value / sky_model.freq_array.si.value) ** 2
+                / (2 * cnst.k_B.si.value)
+            )
+        elif sky_model.stokes.unit.is_equivalent(units.K):
+            conversion = sky_model.stokes.unit.to("K")
+        else:
+            raise ValueError(
```
Comment (r-pascua) [inline hera_sim/visibilities/healvis_wrapper.py:112]: Need to add tests for these lines if we're worried about coverage.

File: setup.cfg [hunk 7]
```diff
@@ -75,6 +75,7 @@ tests =
     hera_cal @ git+git://github.com/hera-team/hera_cal
     healvis @ git+git://github.com/rasg-affiliates/healvis
     bda @ git+git://github.com/HERA-Team/baseline_dependent_averaging.git
+    vis_cpu >= 0.2.0
```
Comment (r-pascua) [inline setup.cfg:78]: Conflict between vis_cpu version listed here and the version listed in the `install_requires` section.

File: hera_sim/visibilities/vis_cpu.py (hunk 3 above)
Comment (piyanatk) [reply hera_sim/visibilities/vis_cpu.py:84]: Yes, I see now that its replace the Numpy docstring "optional" description.

File: setup.cfg [hunk 8]
```diff
@@ -88,6 +89,11 @@ dev =
     healvis @ git+git://github.com/rasg-affiliates/healvis
     hera_cal @ git+git://github.com/hera-team/hera_cal
     bda @ git+git://github.com/HERA-Team/baseline_dependent_averaging.git
+    vis_cpu >= 0.2.0
```
Comment (r-pascua) [inline setup.cfg:92]: Another conflict

File: setup.cfg [hunk 9]
```diff
@@ -88,6 +89,11 @@ dev =
...
+vis =
+    vis_cpu>=0.2.0
```
Comment (r-pascua) [inline setup.cfg:94]: and another

File: hera_sim/tests/test_vis.py [hunk 10]
```diff
@@ -339,429 +459,189 @@ def align_src_to_healpix(point_source_pos, point_source_flux, nside=2 ** 4):
...
+    vis_cpu.HAVE_GPU = old
 
 
-def test_comparison_multiple(uvdata2):
-    freqs = np.unique(uvdata2.freq_array)
+@pytest.mark.parametrize(
+    "simulator, order, conj",
+    product(
+        SIMULATORS, ["time", "baseline", "ant1", "ant2"], ["ant1<ant2", "ant2<ant1"]
+    ),
```
Comment (r-pascua) [inline hera_sim/tests/test_vis.py:530]: Does this take the outer product of the inputs?

File: hera_sim/tests/test_vis.py [hunk 11]
```diff
@@ -339,429 +459,189 @@ def align_src_to_healpix(point_source_pos, point_source_flux, nside=2 ** 4):
...
+    sim.uvdata.reorder_blts(order="time", conj_convention="ant1<ant2")
 
-    healvis = HealVis(
-        uvdata=uvdata2,
-        sky_freqs=freqs,
-        point_source_flux=point_source_flux,
-        point_source_pos=point_source_pos,
-        nside=2 ** 4,
-    ).simulate()
+    print(sim.uvdata.data_array[sim.uvdata.antpair2ind(0, 1)].shape)
```
Comment (r-pascua) [inline hera_sim/tests/test_vis.py:555]: Do we still need the `print` statement here?

File: hera_sim/tests/test_vis.py (hunk 10 above)
Comment (steven-murray) [reply hera_sim/tests/test_vis.py:530]: Yes.

**Comment by r-pascua** [comment]
Oh, I totally forgot that I was supposed to review #170&mdash;I'll get to that ASAP. I think the demo notebook should be relatively painless now that we've fixed all the things that made it kind of a pain in the first place, so I can jump on that too. I'd really like us to start pushing hard on the validation tasks.

File: hera_sim/tests/test_vis.py (hunk 10 above)
Comment (r-pascua) [reply hera_sim/tests/test_vis.py:530]: Neat. Is this syntax preferred to just using multiple `@pytest.parametrize` decorators? If so, then there are some test functions I've written that I should update in the future.

File: hera_sim/visibilities/__init__.py [hunk 12]
```diff
@@ -8,15 +8,13 @@
...
-)
+from .simulators import VisibilitySimulator, VisibilitySimulation, ModelData
 
 # Registered Simulators
-from .vis_cpu import VisCPU
-
+try:
+    from .vis_cpu import VisCPU
+except (ImportError, NameError):
+    pass
```
[the remark this answers is not in the document]
Comment (steven-murray) [reply hera_sim/visibilities/__init__.py:17]: Yup. Good point. Added.

File: hera_sim/visibilities/healvis_wrapper.py (hunk 5 above)
Comment (steven-murray) [reply hera_sim/visibilities/healvis_wrapper.py:62]: Yes, good point. Done.

File: setup.cfg (hunk 8 above)
Comment (steven-murray) [reply setup.cfg:92]: Fixed.

File: hera_sim/visibilities/vis_cpu.py (hunk 2 above)
Comment (piyanatk) [reply hera_sim/visibilities/vis_cpu.py:85]: Is this still need fixing?

File: setup.cfg (hunk 9 above)
Comment (steven-murray) [reply setup.cfg:94]: Fixed.

File: hera_sim/tests/test_vis.py (hunk 11 above)
Comment (steven-murray) [reply hera_sim/tests/test_vis.py:555]: Nope. Removed.

**Review by piyanatk — APPROVED** [review]
Approve!
